mavros/src/plugins/setpoint_trajectory.cpp

```diff
@@ -0,0 +1,180 @@
+
+#include <mavros/mavros_plugin.h>
+#include <mavros/setpoint_mixin.h>
+#include <eigen_conversions/eigen_msg.h>
+
+#include <nav_msgs/Path.h>
+#include <mavros_msgs/PositionTarget.h>
+#include <trajectory_msgs/MultiDOFJointTrajectory.h>
+#include <trajectory_msgs/MultiDOFJointTrajectoryPoint.h>
+
+namespace mavros {
+namespace std_plugins {
+/**
+ * @brief Setpoint TRAJECTORY plugin
+ *
+ * Receive trajectory setpoints and send setpoint_raw setpoints along the trajectory.
+ */
+class SetpointTrajectoryPlugin : public plugin::PluginBase,
+	private plugin::SetPositionTargetLocalNEDMixin<SetpointTrajectoryPlugin> {
+public:
+	SetpointTrajectoryPlugin() : PluginBase(),
+		sp_nh("~setpoint_trajectory")
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		sp_nh.param<std::string>("frame_id", frame_id, "map");
+
+		local_sub = sp_nh.subscribe("local", 10, &SetpointTrajectoryPlugin::local_cb, this);
+		desired_pub = sp_nh.advertise<nav_msgs::Path>("desired", 10);
+
+		sp_timer = sp_nh.createTimer(ros::Duration(0.01), &SetpointTrajectoryPlugin::reference_cb, this, true);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return { /* Rx disabled */ };
+	}
+
+private:
+	friend class SetPositionTargetLocalNEDMixin;
+	ros::NodeHandle sp_nh;
+
+	ros::Timer sp_timer;
+
+	ros::Subscriber local_sub;
+	ros::Publisher desired_pub;
+
+	trajectory_msgs::MultiDOFJointTrajectory::ConstPtr trajectory_target_msg;
+	std::vector<trajectory_msgs::MultiDOFJointTrajectoryPoint>::const_iterator setpoint_target;
+	std::vector<trajectory_msgs::MultiDOFJointTrajectoryPoint>::const_iterator next_setpoint_target;
+
+	std::string frame_id;
+
+	void reset_timer(ros::Duration duration){
+		sp_timer.stop();
+		sp_timer.setPeriod(duration);
+		sp_timer.start();		
+	}
+
+	void publish_path(const trajectory_msgs::MultiDOFJointTrajectory::ConstPtr &req){
+		nav_msgs::Path msg;
+
+		msg.header.stamp = ros::Time::now();
+		msg.header.frame_id = frame_id;
+		for (const auto &p : req->points) {
+			if (p.transforms.empty())
+				continue;
+	
+			geometry_msgs::PoseStamped pose_msg;
+			pose_msg.pose.position.x = p.transforms[0].translation.x;
+			pose_msg.pose.position.y = p.transforms[0].translation.y;
+			pose_msg.pose.position.z = p.transforms[0].translation.z;
+			pose_msg.pose.orientation = p.transforms[0].rotation;
+			msg.poses.emplace_back(pose_msg);
+		}
+		desired_pub.publish(msg);
+	}
+
+	/* -*- callbacks -*- */
+
+	void local_cb(const trajectory_msgs::MultiDOFJointTrajectory::ConstPtr &req)
+	{
+		trajectory_target_msg = req;
+
+		// Read first duration of the setpoint and set the timer
+		setpoint_target = req->points.cbegin();
+		reset_timer(setpoint_target->time_from_start);
+		publish_path(req);
+	}
+
+	void reference_cb(const ros::TimerEvent &event)
+	{
+		using mavlink::common::POSITION_TARGET_TYPEMASK;
+
+		if(!trajectory_target_msg)
+			return;
+
+		Eigen::Vector3d position, velocity, af;
+		Eigen::Quaterniond attitude;
+		float yaw, yaw_rate;
```

Comment:
What about the yaw rate?

Reply:
Good point. Would it make sense to always set it to zero?

Reply:
Why? Can't we handle it as well? You have twist data that you can use

Reply:
That would always enforce the angular rate velocities commands to be valid in the type mask. I am not sure if that is something we want

Reply:
I am not sure I get you: why can't we set the angular rate when the velocity setmask is set for example?

Reply:
I just wasn't sure if we want to use the velocity information in the message. I guess it makes sense to use it actually

Reply:
Done!